### PR TITLE
Fixes STPPaymentCardTextField layout issues on small screens

### PR DESCRIPTION
**Pull request by csabol-stripe**
## Summary
Adds correct formatting for Diner's Club card numbers
When compressed, STPPaymentCardTextField will now use the card number format (not just the fragment length) to decide which numbers to show, matching web behavior
Incomplete card numbers will be marked as invalid, matching web behavior
Correctly sizes card number text field when compressed so that it doesn't push other text fields off screen on small devices

## Motivation
https://jira.corp.stripe.com/browse/IOS-600
https://github.com/stripe/stripe-ios/issues/1007

## Testing
Manually tested on iPhone SE 11.4 simulator in portrait and landscape

### CODE REVIEW

**Comment by danj-stripe** [comment]
Looks like the build is failing on the `check_category_linking.rb` script, and you'll need to add some stuff to `STPCategoryLoader`

File: Stripe/STPPaymentCardTextFieldViewModel.m [hunk 1]
```diff
@@ -20,6 +21,38 @@ - (void)setCardNumber:(NSString *)cardNumber {
     _cardNumber = [sanitizedNumber stp_safeSubstringToIndex:maxLength];
 }
 
+- (NSString *)compressedCardNumber {
```
Comment (danj-stripe) [inline Stripe/STPPaymentCardTextFieldViewModel.m]: WDYT about adding some tests for this?

File: Stripe/STPCardValidator+Private.m [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+@implementation STPCardValidator (Private)
+
++ (NSArray<NSNumber *> *)cardNumberFormatForBrand:(STPCardBrand)brand
+{
+    switch (brand) {
+        case STPCardBrandAmex:
+            return @[@4, @6, @5];
+        case STPCardBrandDinersClub:
+            return @[@4, @6, @4];
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Stripe/STPCardValidator+Private.m]: I thought it should until I saw that they were different. I'm not sure on the source for the 2 fragment length. These values are from https://git.corp.stripe.com/stripe-internal/checkout/blob/d9430b90/manhattan/src/scripts/vendor/stripe-js-v3/src/lib/fields/format.js#L19

File: Stripe/STPPaymentCardTextFieldViewModel.m (hunk 1 above)
Comment (csabol-stripe) [reply Stripe/STPPaymentCardTextFieldViewModel.m]: 👍

File: Stripe/STPPaymentCardTextField.m [hunk 3]
```diff
@@ -718,8 +719,28 @@ - (CGFloat)numberFieldFullWidth {
...
+    NSUInteger fragmentLength = [STPCardValidator fragmentLengthForCardBrand:currentBrand];
+
+    NSMutableString *maxCompressedString = [[NSMutableString alloc] init];
+    while (fragmentLength > maxCompressedString.length) {
+        [maxCompressedString appendString:@"8"];
+    }
+    for (NSNumber *segment in cardNumberFormat) {
+        NSUInteger segmentLength = [segment unsignedIntegerValue];
+        while (segmentLength > maxCompressedString.length) {
+            [maxCompressedString appendString:@"8"];
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Stripe/STPPaymentCardTextField.m]: yeah, let me see if I can simplify

File: Stripe/STPCardValidator+Private.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Stripe/STPCardValidator+Private.m]: Web dashboard shows the last 4
image

**Comment by csabol-stripe** [comment]
@danj-stripe not sure I'm understanding the CI failure here. Do you think it's because I moved that test out of the warning suppression?

File: Stripe/STPCardValidator+Private.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (danj-stripe) [reply Stripe/STPCardValidator+Private.m]: I also do not know the source of the `2`, I think it's (probably?) desirable to update the behavior of `fragmentLengthForCardBrand:`

File: Stripe/STPPaymentCardTextField.m [hunk 4]
```diff
@@ -718,8 +719,23 @@ - (CGFloat)numberFieldFullWidth {
...
+    }
+
+    STPCardBrand currentBrand = [STPCardValidator brandForNumber:cardNumber];
+    NSArray<NSNumber *> *sortedCardNumberFormat = [[STPCardValidator cardNumberFormatForBrand:currentBrand] sortedArrayUsingSelector:@selector(unsignedIntegerValue)];
+    NSUInteger fragmentLength = [STPCardValidator fragmentLengthForCardBrand:currentBrand];
+    NSUInteger maxLength = MAX([[sortedCardNumberFormat lastObject] unsignedIntegerValue], fragmentLength);
+
+    NSMutableString *maxCompressedString = [[NSMutableString alloc] initWithCapacity:maxLength];
+    for (NSUInteger i = 0; i < maxLength; ++i) {
+        [maxCompressedString appendString:@"8"];
```
Comment (danj-stripe) [inline Stripe/STPPaymentCardTextField.m]: Looks like you can do this in a one-liner with

```objective-c
NSString *maxCompressedString = [@"" stringByPaddingToLength:maxLength withString:@"8" startingAtIndex:0];
```

I had no idea that method existed, and this is *totally* fine with me as-is. I was curious if foundation provided something we could use and now that I've found it, wanted to share.

Comment (csabol-stripe) [reply Stripe/STPPaymentCardTextField.m]: oh same. Cool!

**Comment by csabol-stripe** [comment]
Thanks for your help on the CI issues, @danj-stripe !

File: Stripe/STPFormTextField.m [hunk 5]
```diff
@@ -140,7 +140,7 @@ - (void)setAutoFormattingBehavior:(STPFormTextFieldAutoFormattingBehavior)autoFo
                 for (NSNumber *segmentLength in cardNumberFormat) {
                     NSUInteger segmentIndex = 0;
                     for (; index < attributedString.length && segmentIndex < [segmentLength unsignedIntegerValue]; index++, segmentIndex++) {
-                        if (segmentIndex + 1 == [segmentLength unsignedIntegerValue]) {
+                        if (index + 1 != attributedString.length && segmentIndex + 1 == [segmentLength unsignedIntegerValue]) {
```
Comment (danj-stripe) [inline Stripe/STPFormTextField.m]: Ah, nice. This fixed the failing tests for the `intrinsicContentSize`?

Comment (csabol-stripe) [reply Stripe/STPFormTextField.m]: yeah and the extra space at the end of the card number

**Pull request merged**